miku-server/src/main/scala/org/http4s/miku/NettyModelConversion.scala

```diff
@@ -0,0 +1,311 @@
+package org.http4s.miku
+
+import java.net.InetSocketAddress
+
+import cats.effect.{Async, Effect, IO}
+import cats.implicits.{catsSyntaxEither => _, _}
+import com.typesafe.netty.http._
+import fs2.interop.reactivestreams._
+import fs2.{Chunk, Pull, Stream}
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel.Channel
+import io.netty.handler.codec.http._
+import io.netty.handler.codec.http.websocketx.{WebSocketFrame => WSFrame, _}
+import io.netty.handler.ssl.SslHandler
+import org.http4s.Request.Connection
+import org.http4s.headers.`Content-Length`
+import org.http4s.server.websocket.websocketKey
+import org.http4s.util.execution.trampoline
+import org.http4s.websocket.WebSocketContext
+import org.http4s.websocket.WebsocketBits._
+import org.http4s.{HttpVersion => HV, _}
+import org.reactivestreams.{Processor, Subscriber, Subscription}
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.ExecutionContext
+
+/** Helpers for converting http4s request/response
+  * objects to and from the netty model
+  *
+  * Adapted from NettyModelConversion.scala
+  * in
+  * https://github.com/playframework/playframework/blob/master/framework/src/play-netty-server
+  *
+  */
+object NettyModelConversion {
+
+  /** Turn a netty http request into an http4s request
+    *
+    * @param channel the netty channel
+    * @param request the netty http request impl
+    * @return Http4s request
+    */
+  def fromNettyRequest[F[_]](channel: Channel, request: HttpRequest)(
+      implicit F: Effect[F]
+  ): F[Request[F]] = {
+    val connection = createRemoteConnection(channel)
+    if (request.decoderResult().isFailure)
+      F.raiseError(ParseFailure("Malformed request", "Netty codec parsing unsuccessful"))
+    else {
+      val requestBody = convertRequestBody(request)
+      val uri: ParseResult[Uri] = Uri.fromString(request.uri())
+      val headerBuf = new ListBuffer[Header]
+      val headersIterator = request.headers().iteratorAsString()
+      var mapEntry: java.util.Map.Entry[String, String] = null
+      while (headersIterator.hasNext) {
+        mapEntry = headersIterator.next()
+        headerBuf += Header(mapEntry.getKey, mapEntry.getValue)
+      }
+
+      val method: ParseResult[Method] =
+        Method.fromString(request.method().name())
+      val version: HV = {
+        if (request.protocolVersion() == HttpVersion.HTTP_1_1)
+          HV.`HTTP/1.1`
+        else
+          HV.`HTTP/1.0`
```

Comment:
You've covered 99.99999999% of people not causing trouble, but, is this right?

Reply:
This codec can't even respond to http 2.0 messages if you wanted it to. It doesn't handle them properly. I can add a test though

Reply:
To @rossabaker's point, there is nothing restricting the Netty HTTP verison to 1.0 and 1.1:
```
scala> io.netty.handler.codec.http.HttpVersion.valueOf("HTTP/1.2")
res1: io.netty.handler.codec.http.HttpVersion = HTTP/1.2
```

I can't remember where, but I do recall seeing a HTTP/2.0 floating around at $work.

Reply:
Yes, that's what I was thinking.  I think you'd respond correctly, for free.  It's just about honesty in what we pass to the server.  (This is a really, really stupid use case.  It just caught my eye.)

More realistic, though slightly less in 2018: if the service generates an HTTP/1.0 response, I bet this server is supporting some invalid things.  Blaze is pretty careful about the distinction.

Reply:
As a counter chess move, I will point out that http2 has [its own module](https://github.com/netty/netty/tree/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2).

I don't think the `ResponseEncoder` will do valid things for the Http2 spec.